From b3c944c6cfce338e79508dede7ebee0f45b600c8 Mon Sep 17 00:00:00 2001
From: Konstantin Porotchkin <kostap@marvell.com>
Date: Thu, 11 Aug 2016 14:18:33 +0300
Subject: [PATCH 0670/2241] mmc: core: Add MMC_CAP2_NO_DDR50_TUNING to skip SD
 DDR50 tuning.

According to Table 4-22 on Page 73 in SD spec v3.01, tuning CMD19
is only available for SDR50 and SDR104.
Thus some controllers, such as Xenon SDHC, doesn't support tuning
in DDR50.
Set the flag MMC_CAP2_NO_DDR50_TUNING to skip tuning in SD DDR50.

Change-Id: I75ba77924b63868f8818b98a0edbe2a65748c29a
Signed-off-by: Hu Ziji <huziji@marvell.com>
Signed-off-by: Konstantin Porotchkin <kostap@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/31891
Reviewed-by: Hanna Hawa <hannah@marvell.com>
Reviewed-by: Omri Itach <omrii@marvell.com>
Tested-by: Star_Automation <star@marvell.com>
---
 drivers/mmc/core/sd.c    | 3 ++-
 include/linux/mmc/host.h | 1 +
 2 files changed, 3 insertions(+), 1 deletion(-)

diff --git a/drivers/mmc/core/sd.c b/drivers/mmc/core/sd.c
index 967535d..183f5a7 100644
--- a/drivers/mmc/core/sd.c
+++ b/drivers/mmc/core/sd.c
@@ -627,7 +627,8 @@ static int mmc_sd_init_uhs_card(struct mmc_card *card)
 	 */
 	if (!mmc_host_is_spi(card->host) &&
 		(card->host->ios.timing == MMC_TIMING_UHS_SDR50 ||
-		 card->host->ios.timing == MMC_TIMING_UHS_DDR50 ||
+		 (!(card->host->caps2 & MMC_CAP2_NO_DDR50_TUNING) &&
+		 (card->host->ios.timing == MMC_TIMING_UHS_DDR50)) ||
 		 card->host->ios.timing == MMC_TIMING_UHS_SDR104)) {
 		err = mmc_execute_tuning(card);
 
diff --git a/include/linux/mmc/host.h b/include/linux/mmc/host.h
index 8673ffe..00be70e 100644
--- a/include/linux/mmc/host.h
+++ b/include/linux/mmc/host.h
@@ -289,6 +289,7 @@ struct mmc_host {
 #define MMC_CAP2_HSX00_1_2V	(MMC_CAP2_HS200_1_2V_SDR | MMC_CAP2_HS400_1_2V)
 #define MMC_CAP2_SDIO_IRQ_NOTHREAD (1 << 17)
 #define MMC_CAP2_NO_WRITE_PROTECT (1 << 18)	/* No physical write protect pin, assume that card is always read-write */
+#define MMC_CAP2_NO_DDR50_TUNING (1 << 20)	/* Do not execute tuning in DDR50 mode */
 
 	mmc_pm_flag_t		pm_caps;	/* supported pm features */
 
-- 
2.7.4

